ECOM-1339 Branding API footer

wedaly opened the pull request:
Serve branded footer JSON/HTML/CSS/JS from an API endpoint
in the branding app.  Refactor OpenEdX and EdX.org footer templates
to use the Python version of the API, ensuring that the API
values are consistent with the footer included in main.html.

Detailed changes:

* Added footer API end-point to the branding app.
* Footer API allows the language to be set with querystring parameters.
* Footer API allows showing/hiding of the OpenEdX logo using querystring parameters.
* Deprecate ENABLE_FOOTER_V3 in favor of the branding API configuration flag.
* Move no referrer script into main.html from the edx footer template.
* Rename rwd_header_footer.js to rwd_header.js
* Cache API responses.

Authors: Awais Qureshi, Aamir Khan, Will Daly

JIRA: [ECOM-1339](https://openedx.atlassian.net/browse/ECOM-1339)
@jimabramson @AlasdairSwan please review.
FYI: @awais786 @aamir-khan @singingwolfboy @feanil

AlasdairSwan commented on lms/envs/common.py:
```diff
@@ -1827,7 +1872,7 @@
         # translate this the way that Facebook advertises in your language.
         "title": _("Facebook"),
-        "icon": "fa-facebook-square"
+        "icon": "fa-facebook"
```
Marketing want this to be fa-facebook-square

AlasdairSwan commented on lms/envs/common.py:
```diff
@@ -1839,19 +1884,19 @@
         # translate this the way that LinkedIn advertises in your language.
         "title": _("LinkedIn"),
-        "icon": "fa-linkedin-square"
+        "icon": "fa-linkedin"
```
Marketing want this to be fa-linkedin-square

AlasdairSwan commented on lms/envs/common.py:
```diff
@@ -1839,19 +1884,19 @@
         # translate this the way that Google+ advertises in your language.
         "title": _("Google+"),
-        "icon": "fa-google-plus-square"
+        "icon": "fa-google-plus"
```
Marketing want this to be fa-google-plus-square

AlasdairSwan commented on lms/static/js/footer-edx.js:
```diff
@@ -27,7 +15,7 @@ var edx = edx || {};
                      *  to add footer to is in the DOM
                      */
-                    if ( _fn.$el.length > -1 ) {
+                    if ( _fn.$el.length > 0 ) {
```
Actually you just need 
```
if ( _fn.$el.length ) {
```

AlasdairSwan commented on lms/static/sass/_build-lms.scss:
```diff
@@ -21,7 +21,7 @@
 @import 'shared/forms';
 @import 'shared/footer';
-@import 'shared/footer-edx'; // Replaces most of the footer partial. Will update footer to remove edx specific styles once feature flag removed.
+@import 'shared/footer-edx';
```
Shouldn't we keep the comment?

wedaly replied:
I added another comment directly in `shared/footer` next to the styles we want to remove once V3 is turned on.

AlasdairSwan commented:
Some minor changes but once updated and tests pass :+1:

jimabramson commented on lms/djangoapps/branding/api.py:
```diff
@@ -0,0 +1,289 @@
+    # Please do not translate any of these trademarks and company names.
+    return _(
+        u"\u00A9 {org_name}.  All rights reserved except where noted.  "
+        u"EdX, Open edX and the edX and OpenEdX logos are registered trademarks "
```
"Open edX" but not "OpenEdX" ?

AlasdairSwan commented on lms/templates/footer-edx-v3.html:
```diff
@@ -1,90 +1,76 @@
+      <div class="social-media-links">
+        % for link in footer['social_links']:
+        <a href="${link['url']}" class="sm-link external" title="${link['title']}" rel="noreferrer">
+          <span class="icon fa ${link['icon-class']} element-invisible"></span>
```
Sorry @wedaly, this was my mistake. Can you please remove element-invisible and set aria-hidden="true" instead?

jimabramson commented on lms/djangoapps/branding/api.py:
```diff
@@ -0,0 +1,289 @@
+
+    """
+    links = []
+    for social_name in settings.SOCIAL_MEDIA_FOOTER_NAMES:
```
this may be beyond the scope of this PR but is there a reason why the names, urls, and display attributes are stored in separate configuration settings?  It seems somewhat fragile, while using a single dictionary might alleviate the need to fall back on empty defaults for hrefs, css class names, etc

wedaly replied:
I agree that we can simplify this configuration.  Unfortunately, we can't include the "display name" in configuration that gets overridden by Ansible scripts, since the display names are internationalized.  I'm thinking maybe we could simplify it like this:

1. Change `SOCIAL_MEDIA_FOOTER_DISPLAY` from a setting to a constant in `branding.api`.  It's not something we can override in Ansible, so it probably shouldn't be a setting to begin with.

2. Combine `SOCIAL_MEDIA_FOOTER_NAMES` and `SOCIAL_MEDIA_URLS` into a single setting (list of dictionaries with keys `name` and `url`).

Does that make sense?

jimabramson replied:
cool, yes, those sound like good changes.

wedaly replied:
Looking through the code a bit more, I think this change is going to be easier when we remove the V2 version of the footer (which is also using those settings).  That also reduces the risk of deploying this, since unfortunately there are a number of other configuration changes we need to coordinate with this release.

jimabramson replied:
agreed, makes sense to wait until then.

jimabramson commented on lms/djangoapps/branding/api.py:
```diff
@@ -0,0 +1,289 @@
+    logo_name = (
+        u"images/edx-theme/edx-header-logo.png"
+        if settings.FEATURES.get('IS_EDX_DOMAIN', False)
+        else settings.FOOTER_ORGANIZATION_IMAGE
```
does this need to be microsites-aware ?

jimabramson replied:
i'm probably missing something, but if the logo can be customized through this named setting, why do we still need the `IS_EDX_DOMAIN` check?

wedaly replied:
Looks like some leftover cruft from how these templates evolved.  I'll update it to just use the setting (and to be microsites aware).

jimabramson commented on lms/djangoapps/branding/api.py:
```diff
@@ -0,0 +1,289 @@
+    """
+    protocol = "https://" if is_secure else "http://"
+    site_name = microsite.get_value('SITE_NAME', settings.SITE_NAME)
+    return u"{protocol}{site_name}{url_path}".format(
```
i think we should use the [urlparse module](https://docs.python.org/2/library/urlparse.html) for manipulations like this, unless there's a good reason not to.

jimabramson commented on lms/djangoapps/branding/tests/test_views.py:
```diff
@@ -0,0 +1,214 @@
+
+    def test_no_supported_accept_type(self):
+        self._set_feature_flag(True)
+        resp = self._get_footer(accepts="application/x-shockwave-flash")
```
aw.   :-)

jimabramson commented on lms/envs/common.py:
```diff
@@ -1039,6 +1041,48 @@
+FOOTER_CACHE_TIMEOUT = 5 * 60
+
+# Max age cache control header for the footer (controls browser caching).
+FOOTER_BROWSER_CACHE_MAX_AGE = 5 * 60
```
is this the recommended production default?

wedaly replied:
@feanil What do you think?  These settings control the server-side and browser caching of the footer accessed through an API end-point.

feanil replied:
Is there a reason to have a timer on this?  How often would this be invalidated?  Also how would it be invalidated, just by deploys?  Or also some other way?  If this footer can only be changed by deploys, I would up this to at least 30 minutes by default.  Just to make sure, will it save a copy for each language in memcache?

jimabramson replied:
@feanil yes, it gets cached on the server separately per language.

To me it seems it should be safe to up the server side cache (`FOOTER_CACHE_TIMEOUT`) considerably (considering deploys will invalidate that), but the browser-side setting (`FOOTER_BROWSER_CACHE_MAX_AGE`) should probably not exceed a few minutes.

wedaly replied:
@feanil @jimabramson In order for deploys to invalidate the cache, do I need to include the git hash in the cache key?  Or do we have some other way of clearing the cache during a deploy?

jimabramson commented on lms/djangoapps/branding/views.py:
```diff
@@ -102,3 +114,175 @@ def courses(request):
+    # Override the language if necessary
+    language = request.GET.get('language', translation.get_language())
+
+    with translation.override(language):
```
optional: move this call to just above the calls on L270 / L283 and unindent everything a level.  I think this would make the code read a bit clearer, even if it means duplicating a line once.

wedaly commented:
@jimabramson I think I've addressed all your comments.  Could you please take another look at this PR?

@feanil confirmed yesterday that we aren't automatically clearing the default cache in deploys and recommended a 30 minute server-side cache timeout.

wedaly commented on lms/urls.py:
```diff
@@ -427,6 +427,8 @@
         url(r'^courses/{}/edxnotes'.format(settings.COURSE_ID_PATTERN),
             include('edxnotes.urls'), name="edxnotes_endpoints"),
+
+        url(r'^api/v1/branding/', include('branding.api_urls')),
```
I just realized that this URL is not consistent with our API conventions.  It should be `/api/branding/v1/` (version at the end).

@AlasdairSwan I'm going to change this URL, so we'll need to update the URL on Drupal as well.  Can't believe I didn't notice this until just now :\

jimabramson commented:
:+1: looks great @wedaly.   (also, sorry about the incorrect remark regarding cache invalidation)
